virttest/vt_utils/block_disk.py

```diff
@@ -0,0 +1,393 @@
+                         r'(?P<size>\d+\w+)\s+(?P<type>\w+)' % part_num)
+    return pattern.search(process.run(parted_cmd).stdout_text, re.M).groupdict()
+
+
+def _get_mpoint_fstype_linux(partition):
+    """
+    Get a partition's file system type and mountpoint.
+
+    :param partition: disk partition, like /dev/sdb1.
+    :return: Tuple (mountpoint, fstype)
+    """
+    mount_list = process.run('cat /proc/mounts').stdout_text
+    mount_info = re.search(r'%s\s(.+?)\s(.+?)\s' % partition, mount_list)
+    return mount_info.groups()
+
+
+def resize_filesystem_linux(partition, size):
+    """
+    Resize file system.
+    For ext2, ext3, ext4 filesystem, support enlarge and shrink.
+    For xfs filesystem, only support enlarge, not support shrink.
+
+    :param partition: disk partition, like /dev/sdb1.
+    :param size: resize file system to size.
+                 size unit can be 'B', 'K', 'M', 'G'.
+                 support transfer size with SIZE_AVAILABLE,
+                 enlarge to maximun available size.
+    """
+    def get_start_size():
+        start_size = get_partition_attrs_linux(partition)['start']
+        return int(utils_numeric.normalize_data_size(start_size, 'B').split('.')[0])
+
+    def resize_xfs_fs(size):
+        if size == SIZE_AVAILABLE:
+            resize_fs_cmd = 'xfs_growfs -d %s' % mountpoint
+        else:
+            output = process.run('xfs_growfs -n %s' % mountpoint).stdout_text
+            bsize = int(re.findall(r'data\s+=\s+bsize=(\d+)', output, re.M)[0])
+            blocks = (int(utils_numeric.normalize_data_size(size, 'B').split('.')[0]) -
+                      get_start_size()) // bsize
+            resize_fs_cmd = 'xfs_growfs -D %s %s' % (blocks, mountpoint)
+        process.system(resize_fs_cmd)
+
+    def resize_ext_fs(size):
+        flag = False
+        if is_mount(partition, dst=mountpoint, fstype=fstype):
+            umount(partition, mountpoint, fstype=fstype)
+            flag = True
+
+        process.system('e2fsck -f %s' % partition)
+
+        if size == SIZE_AVAILABLE:
+            resize_fs_cmd = 'resize2fs %s' % partition
+        else:
+            output = process.run('tune2fs -l %s | grep -i block' % partition).stdout_text
+            bsize = int(re.findall(r'Block size:\s+(\d+)', output, re.M)[0])
+            size = ((int(utils_numeric.normalize_data_size(size, 'B').split(".")[0]) -
+                     get_start_size()) // bsize) * bsize
+            size = utils_numeric.normalize_data_size(str(size).split(".")[0], 'K')
+            resize_fs_cmd = 'resize2fs %s %sK' % (partition, int(size.split(".")[0]))
+        process.system(resize_fs_cmd)
+        if flag:
+            mount(partition, mountpoint, fstype=fstype)
+
+    mountpoint, fstype = _get_mpoint_fstype_linux(partition)
+    if fstype == 'xfs':
+        resize_xfs_fs(size)
+    elif fstype.startswith("ext"):
+        resize_ext_fs(size)
+    else:
+        raise NotImplementedError
```

Comment:
Suggest adding the error message: NotImplementedError("xxxxxxx")

Reply:
> Suggest adding the error message: NotImplementedError("xxxxxxx")

Done.